<<load-up,echo=FALSE,results=hide>>=
program.name<-'bamchop';
author<-'Zhe Zhang';
affiliation<-'Center for Biomedical Informatics at CHOP';
url<-'https://github.com/CBMi-BiG/bamchop';
@

\documentclass{article}

%headers and footers
\usepackage{fancyhdr}
\usepackage{array}
\pagestyle{fancyplain}

\usepackage{sidecap}
\usepackage{subfigure}
\usepackage{float}
\usepackage{longtable}
\usepackage{colortbl}
\usepackage{setspace}

\fancyhfoffset[LE, LO, RE, RO]{.1in}

\rhead{$\Sexpr{program.name}$}
\chead{}
\lhead{}

\lfoot{\includegraphics[width=0.1\textwidth]{../../logo.jpg} {\bfseries \Sexpr{affiliation}}}
\cfoot{}
\rfoot{\fancyplain{}{\thepage}}

\renewcommand{\footrulewidth}{0.4pt}

\addtolength{\oddsidemargin}{-1in}
\addtolength{\evensidemargin}{-1in}
\addtolength{\textwidth}{2in}
\addtolength{\topmargin}{-0.5in}
\addtolength{\textheight}{1in}

%hyperlink setup
\usepackage{hyperref}
\usepackage{xcolor}
\usepackage{underscore}
\usepackage{appendix}
\definecolor{dark-red}{rgb}{0.4,0.15,0.15}
\definecolor{dark-blue}{rgb}{0.15,0.15,0.4}
\definecolor{medium-blue}{rgb}{0,0,0.5}
\hypersetup{
    colorlinks, linkcolor={dark-red},
    citecolor={dark-blue}, urlcolor={medium-blue}
}

\setlength{\headheight}{15pt}
\setlength{\parindent}{0pt} 
\setlength{\parskip}{1ex}
\setlength{\tabcolsep}{10pt}
\setlength{\extrarowheight}{1.5pt}

\begin{document}
\begin{titlepage}
\title{An introduction to \textit{\Sexpr{program.name}}}
\author{\Sexpr{author}}
\maketitle
\tableofcontents
\end{titlepage}

\section{Installation}

\textit{\Sexpr{program.name}} is based upon \textbf{Sweave} (\url{http://en.wikipedia.org/wiki/Sweave}) framework, which is an intergration of \textbf{R} coding and \textbf{LaTeX} documenting. Therefore, \textit{\Sexpr{program.name}} requires the installation of:

\begin{description}
\item[LaTeX:] Download from \url{http://www.latex-project.org/ftp.html} and install it. The size of the full Mac distribution is about 1.8 GB. You can also use smaller version such as MacTeXtras from \url{http://mirror.ctan.org/systems/mac/mactex/MacTeXtras.zip}. 
\item[R:] \textit{\Sexpr{program.name}} runs within R, so you need to download the newest version of R from \url{http://www.r-project.org} and install it. Bamchop has been tested in R version 2.13 and 2.15. Version 3.01 and above is currently not recommended due to package compliance. The basic functions of R are extended by thousands of packages and \textit{\Sexpr{program.name}} requires the following packages to run:
  \begin{itemize}
  \item \textbf{R.oo}
  \item \textbf{gplots}
  \item \textbf{plotrix}
  \item \textbf{vcd}
  \item \textbf{xtable}
	\end{itemize}
\item[Bioconductor:] Bioconductor is a project extending basic R with hundreds of bioinformatics-related packages. You can install the Bioconductor core packages within R. \textit{\Sexpr{program.name}}  for  to run, plus the following packages:
  \begin{itemize}
  \item \textbf{GenomicRanges}
  \item \textbf{Rsamtools}
  \item \textbf{seqLogo}
  \item \textbf{BSgenome}
  \item reference genome, such as \textbf{BSgenome.Hsapiens.UCSC.hg19} package if hg19 was used for alignment
  \end{itemize}
\item[Bamchop:] Download bamchop from \url{\Sexpr{url}}, and copy the whole folder to the location where will be your local home of bamchop. Replace the logo.jpg file with the logo of your own institute/group. There are a couple of examples in /reports. This subfolder will also be the place to put the results of your own runs.
\end{description}



\subsection{Install R packages within R}
To install an R package within R:
\begin{description}
\item[] Install a specific R package, such as \textbf{vcd}:
<<install1, eval=FALSE>>=
install.packages('vcd', repos="http://lib.stat.cmu.edu/R/CRAN/");
@

\item[] Install Bioconductor core packages:
<<install2, eval=FALSE>>=
source("http://bioconductor.org/biocLite.R");
biocLite();
@

\item[] Install a specific Bioconductor package, such as \textbf{Rsamtools}:
<<install3, eval=FALSE>>=
source("http://bioconductor.org/biocLite.R");
biocLite("Rsamtools");
@
\end{description}
Alternatively, you can download a package to local disk and install it using shell command \textbf{R CMD INSTALL}.


\section{Inputs}
Open the bamchop.r file to edit input variables. The file was split into several sections:
\begin{description}
\item[Mandatory variables. ] These are required variables for each bamchop run. A value must be assigned to each of them. 
\begin{itemize}
  \item \textbf{bamchop.path: } the full path to the bamchop home folder.
  \item \textbf{genome.name: } name of the reference genome used to generate the BAM files. Currently only support hg19, mm9, and ce10.
  \item \textbf{bamflies: } assign the full path of each BAM file to a unique ID. The ID will be used to name the output files.
  \end{itemize}
\item[Extra section:] bamchop can be potentially expanded to describe application-specific information, such as ChIP-seq and targeted resequencing. This feature is not currently supported by bamchop1.01. So no inputs are required in this section. 
\item[Extra parameters:] extra variables that further describe the data set. 
  \begin{itemize}
  \item \textbf{project.name: } name of the data set or project.
  \item \textbf{prepared.by: } the person who made the run. 
  \item \textbf{affiliation: } affiliation of the person.
  \item \textbf{genome.name: } Name of the reference genome.
  \item \textbf{logo: } name of the file with an image to be put in the footer with affiliation. Copy the file into the program folder if you want to use your own.
  \item The rest of the variables in this section are for advanced users. Changes are not recommended.
  \end{itemize}
\end{description}

\section{Run \Sexpr{program.name}}
We are finally ready to run \textit{\Sexpr{program.name}} now. Copy the program folder to a location you prefer and start R within the folder; then use
<<wd, eval=TRUE>>=
getwd();
@
to make sure you in the right working directory; and run \textit{\Sexpr{program.name}} with
<<run, eval=FALSE>>=
source('bamchop.r');
@
The running will take a while. Wait patiently. 

\section{Outputs}
The outputs of each BAM file will be saved in a subfolder of the /results folder, named by the ID of the BAM file. Each folder includes the following files:
\begin{description}
\item[] Two R objects of processed data imported from the BAM file, one full version and one slim version. These objects can be used later to obtain extra information.
\item[] A set of PDF files, each with a plot describing one aspect of the BAM file
\item[] A summarized report with all the PDF files and their detailed explanation, named as ID.pdf.
\end{description}


\section{More information}
\begin{description}
\item[Publication: ] \url{http://www.biomedcentral.com/bmcbioinformatics/supplements/14/S11}
\item[Source code: ] \url{https://github.com/CBMi-BiG/bamchop}
\item[Discussion forum:] \url{http://www.biostars.org/p/81285/}
\end{description}


\end{document}
